src/ray/object_manager/plasma/store.cc

```diff
@@ -249,185 +173,67 @@ PlasmaError PlasmaStore::CreateObject(const ray::ObjectInfo &object_info,
   if (entry == nullptr) {
     return error;
   }
-  ToPlasmaObject(*entry, result, /* check sealed */ false);
+  entry->ToPlasmaObject(result, /* check sealed */ false);
   // Record that this client is using this object.
   AddToClientObjectIds(object_info.object_id, client);
   return PlasmaError::OK;
 }
 
-void PlasmaStore::RemoveGetRequest(const std::shared_ptr<GetRequest> &get_request) {
-  // Remove the get request from each of the relevant object_get_requests hash
-  // tables if it is present there. It should only be present there if the get
-  // request timed out or if it was issued by a client that has disconnected.
-  for (ObjectID &object_id : get_request->object_ids) {
-    auto object_request_iter = object_get_requests_.find(object_id);
-    if (object_request_iter != object_get_requests_.end()) {
-      auto &get_requests = object_request_iter->second;
-      // Erase get_req from the vector.
-      auto it = std::find(get_requests.begin(), get_requests.end(), get_request);
-      if (it != get_requests.end()) {
-        get_requests.erase(it);
-        // If the vector is empty, remove the object ID from the map.
-        if (get_requests.empty()) {
-          object_get_requests_.erase(object_request_iter);
-        }
-      }
-    }
-  }
-  // Remove the get request.
-  get_request->CancelTimer();
-  get_request->MarkRemoved();
-}
-
-void PlasmaStore::RemoveGetRequestsForClient(const std::shared_ptr<Client> &client) {
-  std::unordered_set<std::shared_ptr<GetRequest>> get_requests_to_remove;
-  for (auto const &pair : object_get_requests_) {
-    for (const auto &get_request : pair.second) {
-      if (get_request->client == client) {
-        get_requests_to_remove.insert(get_request);
-      }
-    }
-  }
-
-  // It shouldn't be possible for a given client to be in the middle of multiple get
-  // requests.
-  RAY_CHECK(get_requests_to_remove.size() <= 1);
-  for (const auto &get_request : get_requests_to_remove) {
-    RemoveGetRequest(get_request);
-  }
-}
-
-void PlasmaStore::ReturnFromGet(const std::shared_ptr<GetRequest> &get_req) {
+void PlasmaStore::ReturnFromGet(const std::shared_ptr<GetRequest> &get_request) {
   // If the get request is already removed, do no-op. This can happen because the boost
   // timer is not atomic. See https://github.com/ray-project/ray/pull/15071.
-  if (get_req->IsRemoved()) {
+  if (get_request->IsRemoved()) {
     return;
   }
 
   // Figure out how many file descriptors we need to send.
   absl::flat_hash_set<MEMFD_TYPE> fds_to_send;
   std::vector<MEMFD_TYPE> store_fds;
   std::vector<int64_t> mmap_sizes;
-  for (const auto &object_id : get_req->object_ids) {
-    const PlasmaObject &object = get_req->objects[object_id];
+  for (const auto &object_id : get_request->object_ids) {
+    /// NOTE:(MissionToMars) Record that the client is using the object.
```

Comment:
i wonder if this will have undesired behavior under pressure. As previously we increase the reference count of the object as soon as the object is sealed. but now it's deferred. This means if eviction policy sneaks in between the object might be evicted before the client being able to get a reference of it.

cc @rkooo567  what's your take?

we might just have to call `AddToClientObjectIds` in GetQueue.ObjectSealed eagerly... which is a bit ugly but unavoidable.

Reply:
yeah, the consideration of this change is to make the code cleaner.
maybe we can pass two callback functions to `GetRequestQueue`, the one is called when any object_id is satisfied, and the other one is called only when all object_ids are satisfied.

Reply:
Yeah this is really important. We should not defer it. Please keep the original behavior!

Reply:
(Also we should really avoid changing any behavior in refactoring. That can cause unexpected regression)

Reply:
Defer removed. Keep the same behavior in refactoring.